core/tx_noncer.go

```diff
@@ -49,5 +57,23 @@ func (txn *txNoncer) get(addr common.Address) uint64 {
 // set inserts a new virtual nonce into the virtual state database to be returned
 // whenever the pool requests it instead of reaching into the real state database.
 func (txn *txNoncer) set(addr common.Address, nonce uint64) {
+	txn.lock.Lock()
+	defer txn.lock.Unlock()
+
+	txn.nonces[addr] = nonce
+}
+
+// compareAndSet inserts or updates a new virtual nonce into the virtual state
+// database if the compare callback is true.
+func (txn *txNoncer) compareAndSet(addr common.Address, nonce uint64, compare func(uint64, uint64) bool) {
```

Comment:
Passing a `compare` method in here seems really convoluted, especially since all instances use the same logic anyway. Let's just rename `compareAndSet` to `setLower`, get rid of `compare` altogether and do the comparison internally.